File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.ts
type pair = {
  start_x: number;
  height: number;
};

function largestRectangleArea(heights: number[]): number {
  let max_area = 0;
  let stack: pair[] = [];

  for (let i = 0; i < heights.length; i++) {
    let start = i;
    while (stack.length > 0 && stack[stack.length - 1].height > heights[i]) {
      const { start_x, height } = stack.pop()!;
      const width = i - start_x;
      const area = height * width;
      max_area = Math.max(max_area, area);
      start = start_x;
    }
    stack.push({ start_x: start, height: heights[i] });
  }

  while (stack.length > 0) {
    const { start_x, height } = stack.pop()!;
    const width = heights.length - start_x;
    const area = height * width;
    max_area = Math.max(max_area, area);
  }

  return max_area;
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack: List[tuple(int)] = [] # (idx,height)
        max_area: int = 0

        for idx, h in enumerate(heights):
            start = idx
            while (len(stack)>0) and stack[-1][1] > h:
                (jdx,height) = stack.pop()
                width = idx - jdx
                cur_area = width * height
                max_area = max(max_area,cur_area) 
                start = jdx
            stack.append((start,h))

        total_w = len(heights)

        for (start, height) in (stack):
            cur_area = height*(total_w-start)
            max_area = max(cur_area, max_area)

        return max_area
